src/go/rpk/pkg/cli/cmd/cluster/config/export.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/api/admin"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/config"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/out"
+)
+
+func exportConfig(
+	file *os.File, schema admin.ConfigSchema, config admin.Config, all bool,
+) (err error) {
+	template := make(map[string]interface{})
+
+	// Present properties in alphabetical order, providing some pseudo-grouping based on common prefixes
+	keys := make([]string, 0, len(schema))
+	for k := range schema {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, name := range keys {
+		meta := schema[name]
+		curValue := config[name]
+		template[name] = curValue
+
+		visibility := meta.Visibility
+
+		if visibility == "deprecated" {
+			continue
+		}
+
+		if meta.Visibility == "tunable" && !all {
+			continue
+		}
+
+		// Preface each property with a descriptive comment
+		var comment_tokens []string = make([]string, 0)
+
+		if meta.Example != "" {
+			comment_tokens = append(comment_tokens, fmt.Sprintf("e.g. '%s'", meta.Example))
+		}
+
+		if meta.NeedsRestart {
+			comment_tokens = append(comment_tokens, "restart required")
+		}
+
+		if meta.Nullable {
+			comment_tokens = append(comment_tokens, "may be nil")
+		}
+
+		comment_details := ""
+		if len(comment_tokens) > 0 {
+			comment_details = fmt.Sprintf(" (%s)", strings.Join(comment_tokens, ", "))
+		}
+		file.Write([]byte(fmt.Sprintf("\n# %s%s\n", meta.Description, comment_details)))
+
+		// Compose a YAML representation of the property: this is
+		// done with simple prints rather than the yaml module, because
+		// in either case we have to carefully format values.
+		if meta.Type == "array" {
+			switch x := curValue.(type) {
+			case nil:
+				file.Write([]byte(fmt.Sprintf("%s: []", name)))
+			case []interface{}:
+				if len(x) > 0 {
+					file.Write([]byte(fmt.Sprintf("%s:\n", name)))
+					for _, v := range x {
+						file.Write([]byte(fmt.Sprintf("    - %v\n", v)))
```

Comment:
Does this run the risk of having a value that breaks yaml?

Reply:
Yes, although not in normal use.  I think someone would have to make an effort to inject invalid yaml.  Then the consequence would be that when saving the config, the invalid yaml would get rejected (only the export part uses prints, the import path still uses proper yaml parsing.  Most of these types of things are probably also the kind of thing we should be improving validation for on the backend (currently the most instances of `property<sstring>` for example are very liberal and would accept something like an s3 auth key with a newline in it)

afact the golang yaml libraries don't provide a helper for sanitizing strings outside of yaml-marshalling the whole lot -- I'm inclined to stay with the print version for its slightly cleaner output (1 blank line between entries instead of 2), but it's easy enough to just drop that commit and go back to using the full-on yaml marshalling for output to remove the risk of non-yaml output.

Reply:
Cool, I like the print version more, I agree with your reasoning 👍